src/resolution/composite-arns-resolver.ts

```diff
@@ -17,33 +17,69 @@
  */
 import winston from 'winston';
 import { NameResolution, NameResolver } from '../types.js';
+import * as metrics from '../metrics.js';
+import { KvArnsStore } from '../store/kv-arns-store.js';
 
 export class CompositeArNSResolver implements NameResolver {
   private log: winston.Logger;
   private resolvers: NameResolver[];
+  private cache: KvArnsStore;
 
   constructor({
     log,
     resolvers,
+    cache,
   }: {
     log: winston.Logger;
     resolvers: NameResolver[];
+    cache: KvArnsStore;
   }) {
-    this.log = log.child({ class: 'CompositeArNSResolver' });
+    this.log = log.child({ class: this.constructor.name });
     this.resolvers = resolvers;
+    this.cache = cache;
   }
 
   async resolve(name: string): Promise<NameResolution> {
     this.log.info('Resolving name...', { name });
 
     try {
+      const cachedResolutionBuffer = await this.cache.get(name);
+      if (cachedResolutionBuffer) {
+        const cachedResolution: NameResolution = JSON.parse(
+          cachedResolutionBuffer.toString(),
+        );
+        if (
+          cachedResolution !== undefined &&
+          cachedResolution.resolvedAt !== undefined &&
+          cachedResolution.ttl !== undefined &&
+          cachedResolution.resolvedAt + cachedResolution.ttl * 1000 > Date.now()
+        ) {
+          metrics.arnsCacheHitCounter.inc();
+          this.log.info('Cache hit for arns name', { name });
+          return cachedResolution;
+        }
+      }
+      metrics.arnsCacheMissCounter.inc();
+      this.log.info('Cache miss for arns name', { name });
+
       for (const resolver of this.resolvers) {
-        this.log.debug('Attempting to resolve name with resolver', {
-          resolver,
-        });
-        const resolution = await resolver.resolve(name);
-        if (resolution.resolvedId !== undefined) {
-          return resolution;
+        try {
+          this.log.info('Attempting to resolve name with resolver', {
+            type: resolver.constructor.name,
+            name,
+          });
+          const resolution = await resolver.resolve(name);
+          if (resolution.resolvedId !== undefined) {
+            this.cache.set(name, Buffer.from(JSON.stringify(resolution)));
+            this.log.info('Resolved name', { name, resolution });
+            return resolution;
+          }
+        } catch (error: any) {
+          this.log.error('Error resolving name with resolver', {
+            resolver,
+            message: error.message,
+            stack: error.stack,
+          });
         }
       }
```

Comment:
Should we fall back to the cached resolution here, perhaps with some staleness threshold, if we have one?

Reply:
modified to return the cached resolution data on error, if we have it here - [b9e4fe6](https://github.com/ar-io/ar-io-node/pull/200/commits/b9e4fe63b1e5d876ee5078577c15d709f5a8ffcd)